src/odemis/driver/smaract.py

```diff
@@ -1747,27 +1759,101 @@ def SA_MC_Move(self, id, p_pose):
         self.stopping.clear()
         pose = _deref(p_pose, SA_MC_Pose)
         if self._pose_in_range(pose):
-            self._current_move_finish = time.time() + 1.0
             self.target.x = pose.x
             self.target.y = pose.y
             self.target.z = pose.z
             self.target.rx = pose.rx
             self.target.ry = pose.ry
             self.target.rz = pose.rz
 
-            logging.debug("sim MC5DOF: moving to target: %s" % (self.target,))
+            lin_speed = self.properties[MC_5DOF_DLL.SA_MC_PKEY_MAX_SPEED_LINEAR_AXES].value
+            rad_speed = self.properties[MC_5DOF_DLL.SA_MC_PKEY_MAX_SPEED_ROTARY_AXES].value
+
+            # estimate move duration
+            dur = max(
+                abs(self.target.x - self.pose.x) / lin_speed,
+                abs(self.target.y - self.pose.y) / lin_speed,
+                abs(self.target.z - self.pose.z) / lin_speed,
+                abs(self.target.rx - self.pose.rx) / rad_speed,
+                abs(self.target.rz - self.pose.rz) / rad_speed,
+                )
+
+            self._current_move_finish = time.time() + dur
+            self._last_time = time.time()
+            logging.debug("sim MC5DOF: moving to target: %s duration %f s" % (self.target, dur))
         else:
             raise SA_MCError(MC_5DOF_DLL.SA_MC_ERROR_POSE_UNREACHABLE, "error")
 
     def SA_MC_GetPose(self, id, p_pose):
         pose = _deref(p_pose, SA_MC_Pose)
+
+        cur_time = time.time()
+        if cur_time < self._current_move_finish:
+            lin_speed = self.properties[MC_5DOF_DLL.SA_MC_PKEY_MAX_SPEED_LINEAR_AXES].value
+            rad_speed = self.properties[MC_5DOF_DLL.SA_MC_PKEY_MAX_SPEED_ROTARY_AXES].value
+            dt = cur_time - self._last_time
+
+            # calculate intermediate positions
+            dx = self.target.x - self.pose.x
+            if dx >= 0:
+                self.pose.x = self.pose.x + lin_speed * dt
+                if self.pose.x >= self.target.x:
+                    self.pose.x = self.target.x
+            elif dx < 0:
+                self.pose.x = self.pose.x - lin_speed * dt
+                if self.pose.x < self.target.x:
+                    self.pose.x = self.target.x
```

Comment:
This is 5 times the same code, with slightly different variable names. Please move to a separate function, and call it 5 times.